include/swift/AST/Import.h

```diff
@@ -0,0 +1,350 @@
+    /// variable, this type can be used mid-expression to produce a scope path
+    /// that is valid until the end of the expression.
+    using Builder = detail::ImportPathBuilder<Access>;
+
+    Access(ImportPath::Raw raw) : ImportPathBase(raw) {
+      assert(size() <= 1 && "nested scoped imports are not supported");
+    }
+
+    Access() : ImportPathBase({}) { }
+
+    /// Returns \c true if the scope of this import includes \c name. An empty
+    /// scope matches all names.
+    bool matches(DeclName name) const {
+      return empty() || DeclName(front().Item).matchesRef(name);
+    }
+  };
+
+  /// Represents a module path--the portion of an \c ImportPath which describes
+  /// the name of the module being imported, possibly including submodules.
+  ///
+  /// \c ImportPath::Module contains one or more identifiers. The first
+  /// identiifer names a top-level module. The second and subsequent
+  /// identifiers, if present, chain together to name a specific submodule to
+  /// import. (Although Swift modules cannot currently contain submodules, Swift
+  /// can import Clang submodules.)
+  ///
+  /// \c ImportPath::Module is essentially a wrapper around \c ArrayRef and
+  /// does not own its elements, so something else needs to manage their
+  /// lifetime. \c ImportDecl owns the memory backing
+  /// \c ImportDecl::getModulePath().
+  class Module : public detail::ImportPathBase<Module> {
+  public:
+    /// A helper type which encapsulates a temporary vector and can produce a
+    /// module path from it. In addition to the obvious use in a temporary
+    /// variable, this type can be used mid-expression to produce a module path
+    /// that is valid until the end of the expression.
+    using Builder = detail::ImportPathBuilder<Module>;
+
+    Module(ImportPath::Raw raw) : ImportPathBase(raw) {
+      assert(size() >= 1 && "must have a top-level module");
+    }
+
+    // Note: This type does not have a constructor which just takes an
+    // `Identifier` because it would not be able to create a temporary
+    // `ImportPath::Element` with a long enough lifetime to return. Use
+    // `ImportPath::Module::Builder` to create a temporary module path.
+
+    bool hasSubmodule() const {
+      return size() != 1;
+    }
+
+    ImportPath::Raw getSubmodulePath() const {
+      return getRaw().drop_front();
+    }
+  };
+
+  ImportPath(Raw raw) : ImportPathBase(raw) {
+    assert(raw.size() >= 1 && "ImportPath must contain a module name");
+  }
+
+  /// Extracts the portion of the \c ImportPath which represents a module name,
+  /// including submodules if appropriate.
+  Module getModulePath(bool isScoped) const {
+    if (isScoped)
+      return Module(getRaw().drop_back());
+
+    return Module(getRaw());
+  }
+
+  /// Extracts the portion of the \c ImportPath which represents a scope for the
+  /// import.
+  Access getAccessPath(bool isScoped) const {
+    if (isScoped)
+      return Access(getRaw().take_back());
```

Comment:
Should this assert that there's at least two elements in the path? Just to make sure we don't accidentally pick up a module name

Reply:
Good idea.